### PR TITLE
feat: use `git-version` instead of `git2`

### PR DESCRIPTION
**Pull request by mbrobbel**
An alternative approach for #172. This replaces the `git2` dependency with `git-version` which uses the `git` binary directly at compile time. This means that:

For dev builds, needs `git` binary:
- There is a submodule checked out: version file is written to the `gen` folder.
- There is no submodule checked out: the build fails with a hint to clone the submodule.

For packaged builds (substrait is not included as a git submodule, see `Cargo.toml` includes), don't need a `git` binary:
- The version file is in the `gen` folder: the build script skips the version generation part
- The version file is missing: the build fails - the package is broken

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -10,57 +10,56 @@ use std::{
 };
 use walkdir::{DirEntry, WalkDir};
 
+const SUBMODULE_ROOT: &str = "substrait";
 const PROTO_ROOT: &str = "substrait/proto";
 const TEXT_ROOT: &str = "substrait/text";
+const GEN_ROOT: &str = "gen";
```
Comment (westonpace) [inline build.rs:16]: This is a good idea to mark that the code is generated

**Review by westonpace — APPROVED** [review]
Thanks for looking into this.

**Comment by westonpace** [comment]
Ah, this doesn't seem to be working quite right. I get an error using the released build (because `git_submodule_versions!` runs immediately, being a macro, even if `version.in` exists).  I'll investigate.

```
error: failed to determine .git directory: git rev-parse exited with status 128: fatal: not a git repository (or any of the parent directories): .git
  --> /home/pace/.cargo/registry/src/index.crates.io-6f17d22bba15001f/substrait-0.29.1/build.rs:40:31
   |
40 |           const VERSION: &str = git_version::git_submodule_versions!(
   |  _______________________________^
41 | |             args = ["--tags", "--long", "--dirty=-dirty", "--abbrev=40"],
42 | |             fallback = ""
43 | |         )[0]
   | |_________^
   |
   = note: this error originates in the macro `git_version::git_submodule_versions` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `substrait` (build script) due to previous error
```

**Comment by mbrobbel** [comment]
> Ah, this doesn't seem to be working quite right. I get an error using the released build (because `git_submodule_versions!` runs immediately, being a macro, even if `version.in` exists). I'll investigate.
> 
> ```
> error: failed to determine .git directory: git rev-parse exited with status 128: fatal: not a git repository (or any of the parent directories): .git
>   --> /home/pace/.cargo/registry/src/index.crates.io-6f17d22bba15001f/substrait-0.29.1/build.rs:40:31
>    |
> 40 |           const VERSION: &str = git_version::git_submodule_versions!(
>    |  _______________________________^
> 41 | |             args = ["--tags", "--long", "--dirty=-dirty", "--abbrev=40"],
> 42 | |             fallback = ""
> 43 | |         )[0]
>    | |_________^
>    |
>    = note: this error originates in the macro `git_version::git_submodule_versions` (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> error: could not compile `substrait` (build script) due to previous error
> ```

Hm, it should use the provided `fallback` (when the `git describe` fails), I saw this error while testing, and it went away for me after adding the `fallback`.

**Comment by westonpace** [comment]
Ah, it seems `git_versions!` respects fallback but not `git_submodule_verisons!`

**Comment by westonpace** [comment]
Ah, nevermind, `git_submodule_verisons!` does look at fallback but only for some errors.

**Comment by mbrobbel** [comment]
> Thanks. I think the problem is that, with this kind of failure, `git_submodule_versions` doesn't even know how many tuples to return (because it doesn't know how many submodules there are). Maybe you can add a new functions `git_submodule_version` which requires a submodule name as input, and which always respects `fallback`.

Yes, I was considering the same, but then I figured we can probably be pragmatic here: https://github.com/substrait-io/substrait-rs/pull/175.

**Pull request merged**